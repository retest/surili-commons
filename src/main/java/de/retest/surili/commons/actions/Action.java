package de.retest.surili.commons.actions;

import java.io.Serializable;
import java.util.Optional;

import de.retest.recheck.ui.descriptors.Element;

public interface Action extends Serializable {

	Optional<Element> getTargetElement();

}
