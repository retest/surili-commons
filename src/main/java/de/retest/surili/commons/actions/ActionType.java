package de.retest.surili.commons.actions;

import java.io.Serializable;
import java.util.Collection;
import java.util.Set;
import java.util.stream.Collectors;

import lombok.NonNull;
import lombok.Value;

/**
 * Every {@link Action} instance belongs to a certain action type. This abstraction helps to limit the number of
 * possible action types in a {@code SutState}. Target elements are represented by the unique path (e.g. XPath) of an
 * element.
 */
@Value
public class ActionType implements Serializable {

	private static final long serialVersionUID = 1L;

	public static final String EMPTY_TARGET_ELEMENT_PATH = "";

	@NonNull
	String elementPath;
	@NonNull
	String type;

	public static ActionType fromAction( final Action action ) {
		final String elementPath = action.getTargetElement()
				.map( element -> element.getIdentifyingAttributes().getPath() ).orElse( EMPTY_TARGET_ELEMENT_PATH );
		final String type = action.getClass().getName();
		return new ActionType( elementPath, type );
	}

	public static Set<ActionType> fromCollection( final Collection<Action> actions ) {
		return actions.stream().map( ActionType::fromAction ).collect( Collectors.toSet() );
	}
}
