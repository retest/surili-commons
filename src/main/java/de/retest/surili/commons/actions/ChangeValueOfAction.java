package de.retest.surili.commons.actions;

import java.util.Optional;

import de.retest.recheck.ui.descriptors.Element;
import lombok.NonNull;
import lombok.Value;

@Value
public class ChangeValueOfAction implements Action {

	private static final long serialVersionUID = 1L;

	@NonNull
	Element targetElement;
	@NonNull
	CharSequence[] keysToSend;

	@Override
	public Optional<Element> getTargetElement() {
		return Optional.of( targetElement );
	}

}
