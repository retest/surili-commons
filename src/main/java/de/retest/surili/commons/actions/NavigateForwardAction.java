package de.retest.surili.commons.actions;

import java.util.Optional;

import de.retest.recheck.ui.descriptors.Element;
import lombok.Value;

@Value
public class NavigateForwardAction implements Action {

	private static final long serialVersionUID = 1L;

	@Override
	public Optional<Element> getTargetElement() {
		return Optional.empty();
	}

}
