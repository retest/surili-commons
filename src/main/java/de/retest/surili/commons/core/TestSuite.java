package de.retest.surili.commons.core;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * A test suite with a set (i.e. unordered) of {@link TestCase}s.
 */
public interface TestSuite extends Serializable {

	/**
	 * @return The set of test cases of this test suite.
	 */
	Set<TestCase> getTestCases();

	/**
	 * @return {@link #getTestCases()} as a list.
	 */
	default List<TestCase> getTestCasesAsList() {
		return new ArrayList<>( getTestCases() );
	}

	/**
	 * @return The number of test cases of this test suite.
	 */
	int getSize();

}
