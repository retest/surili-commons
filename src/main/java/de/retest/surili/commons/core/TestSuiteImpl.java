package de.retest.surili.commons.core;

import java.util.Set;

import lombok.NonNull;
import lombok.Value;

@Value
public class TestSuiteImpl implements TestSuite {

	private static final long serialVersionUID = 1L;

	@NonNull
	private final Set<TestCase> testCases;

	@Override
	public int getSize() {
		return testCases.size();
	}

}
