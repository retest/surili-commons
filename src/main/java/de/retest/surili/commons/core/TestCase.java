package de.retest.surili.commons.core;

import java.io.Serializable;
import java.util.List;

import de.retest.surili.commons.actions.Action;

public interface TestCase extends Serializable {

	String getName();

	List<Action> getActions();

}
