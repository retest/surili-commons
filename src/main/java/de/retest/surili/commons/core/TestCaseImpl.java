package de.retest.surili.commons.core;

import java.util.List;
import java.util.Optional;

import de.retest.surili.commons.actions.Action;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.Value;

@Value
@RequiredArgsConstructor
public class TestCaseImpl implements TestCase {

	private static final long serialVersionUID = 1L;

	private final String name;
	@NonNull
	private final List<Action> actions;

	public TestCaseImpl( final List<Action> actions ) {
		this( null, actions );
	}

	@Override
	public Optional<String> getName() {
		return Optional.ofNullable( name );
	}

	@Override
	public int getSize() {
		return actions.size();
	}

}
