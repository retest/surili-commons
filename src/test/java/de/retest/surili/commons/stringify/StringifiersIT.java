package de.retest.surili.commons.stringify;

import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Set;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import de.retest.surili.commons.actions.Action;
import de.retest.surili.commons.actions.NavigateBackAction;
import de.retest.surili.commons.actions.NavigateForwardAction;
import de.retest.surili.commons.core.TestCase;
import de.retest.surili.commons.core.TestCaseImpl;
import de.retest.surili.commons.core.TestSuite;
import de.retest.surili.commons.core.TestSuiteImpl;

class StringifiersIT {

	Action action;
	TestCase testCase;
	TestSuite testSuite;

	@BeforeEach
	void setUp() throws Exception {
		action = new NavigateForwardAction();
		final Action otherAction = new NavigateBackAction();

		testCase = new TestCaseImpl( Arrays.asList( action, otherAction ) );
		final TestCase otherTestCase = new TestCaseImpl( Arrays.asList( otherAction, action ) );

		final Set<TestCase> testCases = new LinkedHashSet<>( Arrays.asList( testCase, otherTestCase ) );

		testSuite = new TestSuiteImpl( testCases );
	}

	@Test
	void should_handle_actions() throws Exception {
		final Stringifier<Action> cut = new ActionStringifier();

		final String stringified = cut.toString( action );

		Approvals.verify( stringified );
	}

	@Test
	void should_handle_test_cases() throws Exception {
		final Stringifier<TestCase> cut = new TestCaseStringifier();

		final String stringified = cut.toString( testCase );

		Approvals.verify( stringified );
	}

	@Test
	void should_handle_test_suites() throws Exception {
		final Stringifier<TestSuite> cut = new TestSuiteStringifier();

		final String stringified = cut.toString( testSuite );

		Approvals.verify( stringified );
	}

	@Test
	void should_handle_empty_test_cases() throws Exception {
		final Stringifier<TestCase> cut = new TestCaseStringifier();

		final String stringified = cut.toString( new TestCaseImpl( Collections.emptyList() ) );

		Approvals.verify( stringified );
	}

	@Test
	void should_handle_empty_test_suites() throws Exception {
		final Stringifier<TestSuite> cut = new TestSuiteStringifier();

		final String stringified = cut.toString( new TestSuiteImpl( Collections.emptySet() ) );

		Approvals.verify( stringified );
	}

	@Test
	void should_handle_empty_test_suites_with_empty_test_cases() throws Exception {
		final TestCase emptyTestCase = new TestCaseImpl( Collections.emptyList() );
		final Set<TestCase> testCases = new LinkedHashSet<>( Arrays.asList( testCase, emptyTestCase ) );
		final TestSuite testSuite = new TestSuiteImpl( testCases );
		final Stringifier<TestSuite> cut = new TestSuiteStringifier();

		final String stringified = cut.toString( testSuite );

		Approvals.verify( stringified );
	}

}
