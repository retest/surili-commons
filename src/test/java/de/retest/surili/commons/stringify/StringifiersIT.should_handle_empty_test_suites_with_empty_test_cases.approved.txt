TestSuite:
	TestCase:
		NavigateForwardAction()
		NavigateBackAction()
	TestCase: empty
